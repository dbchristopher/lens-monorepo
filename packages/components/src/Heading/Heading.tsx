import omit from 'lodash/omit';
import React, { FunctionComponent, Ref } from 'react';
import styled, { StyledComponent } from 'styled-components';
import {
  getTextTransform,
  ResponsiveFontSize,
  shouldTruncate,
  TextTransforms,
  textVariant,
  TextVariants,
} from 'looker-lens-design-tokens';
import { BoxPropsWithout } from '../Box/Box';
import Box from '../Box';

export type HeadingLevels = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
export type HeadingTextTransforms = 'caps' | 'lower' | 'none' | 'upper';

export interface HeadingProps
  extends BoxPropsWithout<HTMLHeadingElement, 'truncate'> {
  /** Heading level from h1-h6 */
  is?: HeadingLevels;
  /** Text tranform  */
  textTransform?: TextTransforms;
  /** Truncate heading text */
  truncate?: boolean;
  /** Truncate heading at a specified number of lines (whole number) */
  truncateLines?: number;
  /** Custom css class */
  className?: string;
  /** Adjust style of text with more meaning by using an intent */
  variant?: TextVariants;
}

type ComponentType = FunctionComponent<HeadingProps>;
type StyledComponentType = StyledComponent<ComponentType, HeadingProps>;

const InternalHeading: ComponentType = ({
  fontSize,
  fontWeight,
  lineHeight,
  is,
  ...props
}) => (
  <Box
    is={is || 'h2'}
    fontSize={fontSize || headingLevelSize(is)}
    lineHeight={lineHeight || headingLineHeight(is, fontSize)}
    fontWeight={fontWeight || 'normal'}
    {...omit(props, ['textTransform', 'truncate', 'truncateLines'])}
  >
    {props.children}
  </Box>
);

const HeadingFactory = React.forwardRef<StyledComponentType, HeadingProps>(
  (props: HeadingProps, ref: Ref<StyledComponentType>) => (
    <InternalHeading ref={ref} {...props} />
  )
);

const Heading = styled<ComponentType>(HeadingFactory)`
  ${props => getTextTransform(props.textTransform)};
  ${props => shouldTruncate(props.truncate, props.truncateLines)};
  ${props => textVariant(props.theme, props.variant)};
`;

export default Heading;

const headingLevelSize = (is?: HeadingLevels) => {
  switch (is) {
    case 'h1':
      return 'xxlarge';
    case 'h2':
      return 'xlarge';
    case 'h3':
      return 'large';
    case 'h4':
      return 'medium';
    case 'h5':
      return 'small';
    case 'h6':
      return 'xsmall';
    default:
      return 'large';
  }
};

const headingLineHeight = (is?: HeadingLevels, size?: ResponsiveFontSize) => {
  if (size) return size;

  switch (is) {
    case 'h1':
      return 'xxlarge';
    case 'h2':
      return 'xlarge';
    case 'h3':
      return 'large';
    case 'h4':
      return 'medium';
    case 'h5':
      return 'small';
    case 'h6':
      return 'xsmall';
    default:
      return 'large';
  }
};
