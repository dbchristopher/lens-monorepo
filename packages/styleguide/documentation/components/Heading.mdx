```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources')
  .StatusAndResources
;<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/SmIM3VUQUgo8xb4o7UUxflkT/Typography"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Heading/Heading.tsx"
  feedbackTitle="Heading Component Feedback"
/>
```

### Heading Default

The `<Heading />` component is used to render a HTML `<h1>` - `<h6>` element, by default it will render a `<h2>` element

```js
// A heading component that defaults to a h3
import { Heading } from './Heading'
;<Heading>Hello Good Looker üëã</Heading>
```

<div class="doc-section-divider"></div>

### Heading Levels

To use a different HTML heading element, the `<Heading />` component accepts a `is` attribute that corresponds to the `<h1>` - `<h6>` elements. The font-size of each heading element maps to the Lens [type ramp](/#!/Typography)

```js
import { Heading } from './Heading'
// Heading components using the level attribute (font-size / line-height)
;<div>
  <Heading is="h1">I‚Äôm a h1 element (25px/40px)</Heading>
  <Heading is="h2">I‚Äôm a h2 element (22px/32px)</Heading>
  <Heading is="h3">I‚Äôm a h3 element (18px/28px)</Heading>
  <Heading is="h4">I‚Äôm a h4 element (16/24px)</Heading>
  <Heading is="h5">I‚Äôm a h5 element (14px/20px)</Heading>
  <Heading is="h6">I‚Äôm a h6 element (12px/16px)</Heading>
</div>
```

<div class="doc-section-divider"></div>

### Heading Sizes

When creating accessible pages it is important that headings create a [logical document outline](https://bitsofco.de/using-heading-elements-to-create-a-document-outline/), but sometimes the font-size of the heading element doesn't match to the needs of the design or layout. Composing the `is` and the `fontSize` attributes lets you choose the semantically correct level heading and the desired size. The available size values come from the [type ramp](/#!/Typography).

```js
import { Heading } from './Heading'
// Heading components using the is and fontSize attribute (font-size / line-height)
;<div>
  <Heading fontSize="xxxxlarge">Viral meditation live-edge</Heading>
  <Heading fontSize="xxxlarge">Pork belly beard mustache </Heading>
  <Heading fontSize="xxlarge">Iceland trust fund hell of plaid</Heading>
  <Heading fontSize="xlarge">Tofu heirloom hammock fam tweet</Heading>
  <Heading fontSize="large">Helvetica hashtag cronut unicorn</Heading>
  <Heading fontSize="medium">
    Franzen semiotics sustainable, gluten-free
  </Heading>
  <Heading fontSize="small">
    8-bit coloring book authentic squid pok pok
  </Heading>
  <Heading fontSize="xsmall">
    Beard truffaut fashion axe, butcher raclette
  </Heading>
  <Heading fontSize="xxsmall">Man bun cardigan succulents vice</Heading>
</div>
```

<div class="doc-section-divider"></div>

### **Weight and Transform**

Another common pattern for headings is to control the font-weight and the text-transform properties. The `<Heading />` component allows you to adjust those with the `fontWweight` and `textTransform` attributes.

```js
import { Heading } from './Heading'
// Heading components using the fontWeight and textTransform attributes
;<>
  <Heading fontWeight="light" textTransform="lower">
    Light and lower
  </Heading>
  <Heading fontWeight="normal" textTransform="none">
    Normal and none (default)
  </Heading>
  <Heading fontWeight="semiBold" textTransform="upper">
    Semi-bold and upper
  </Heading>
  <Heading fontWeight="bold" textTransform="caps">
    Bold and caps
  </Heading>
</>
```

<div class="doc-section-divider"></div>

### Heading alignment

The `align` property allows you to adjust the `text-align` property of your `<Heading />` component. This is useful if you need to center or right align the text.

```js
import { Heading } from './Heading'
;<>
  <Heading align="left">‚óÄÔ∏è Align left (Default) </Heading>
  <Heading align="center">‚óÄÔ∏è Align Center ‚ñ∂Ô∏è</Heading>
  <Heading align="right">Align Right ‚ñ∂Ô∏è</Heading>
</>
```

<div class="doc-section-divider"></div>

### Heading Variants

Lens provides a few variants for the `<Heading />` component based on common patterns we use in our applications, the `variant` property will change that will change the color the rendered heading.

```js
import { Heading } from './Heading'
;<>
  <Heading>Regular Heading</Heading>
  <Heading variant="secondary">Secondary Heading</Heading>
  <Heading variant="subdued">Subdued Heading</Heading>
  <Heading variant="critical">Critical Heading</Heading>
  <Heading variant="positive">Positive Heading</Heading>
</>
```

<div class="doc-section-divider"></div>
