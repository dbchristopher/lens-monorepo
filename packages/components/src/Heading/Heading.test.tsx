import * as React from 'react';
import { createWithTheme } from '../../../test/utils/create_with_theme';
import { Heading } from './Heading';

test('A default Heading', () => {
  const component = createWithTheme(<Heading>ðŸ¥‘</Heading>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A <h1> Heading', () => {
  const component = createWithTheme(<Heading is="h1">ðŸ¥‘</Heading>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A <h1> Heading sized to <h2>', () => {
  const component = createWithTheme(
    <Heading is="h1" fontSize="xlarge">
      ðŸ¥‘
    </Heading>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A Heading to bold', () => {
  const component = createWithTheme(<Heading fontWeight="bold">ðŸ¥‘</Heading>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A Heading transformed', () => {
  const component = createWithTheme(<Heading textTransform="caps">ðŸ¥‘</Heading>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A Heading with variant', () => {
  const component = createWithTheme(<Heading variant="subdued">ðŸ¥‘</Heading>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A Heading with color', () => {
  const component = createWithTheme(
    <Heading color="palette.charcoal100">ðŸ¥‘</Heading>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A Heading truncated', () => {
  const component = createWithTheme(<Heading truncate>ðŸ¥‘</Heading>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('A Heading with multiline truncate', () => {
  const component = createWithTheme(<Heading truncateLines={2}>ðŸ¥‘</Heading>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
