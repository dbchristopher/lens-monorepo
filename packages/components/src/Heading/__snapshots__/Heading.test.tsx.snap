// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <h1> Heading 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.5625rem;
  font-weight: 400;
  line-height: 2.5rem;
}

.c0 {
  text-transform: none;
}

<h1
  className="c0 c1"
>
  ðŸ¥‘
</h1>
`;

exports[`A <h1> Heading sized to <h2> 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.375rem;
  font-weight: 400;
  line-height: 2rem;
}

.c0 {
  text-transform: none;
}

<h1
  className="c0 c1"
>
  ðŸ¥‘
</h1>
`;

exports[`A Heading to bold 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.125rem;
  font-weight: 700;
  line-height: 1.75rem;
}

.c0 {
  text-transform: none;
}

<h2
  className="c0 c1"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading transformed 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1.75rem;
}

.c0 {
  text-transform: capitalize;
}

<h2
  className="c0 c1"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading truncated 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1.75rem;
}

.c0 {
  text-transform: none;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<h2
  className="c0 c1"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading with color 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  color: #F5F6F7;
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1.75rem;
}

.c0 {
  text-transform: none;
}

<h2
  className="c0 c1"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading with multiline truncate 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1.75rem;
}

.c0 {
  text-transform: none;
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 2;
}

<h2
  className="c0 c1"
>
  ðŸ¥‘
</h2>
`;

exports[`A Heading with variant 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1.75rem;
}

.c0 {
  text-transform: none;
  color: #939BA5;
}

<h2
  className="c0 c1"
>
  ðŸ¥‘
</h2>
`;

exports[`A default Heading 1`] = `
.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  font-size: 1.125rem;
  font-weight: 400;
  line-height: 1.75rem;
}

.c0 {
  text-transform: none;
}

<h2
  className="c0 c1"
>
  ðŸ¥‘
</h2>
`;
